monkey/agent_plugins/exploiters/smb/src/smb_exploit_client.py

```diff
@@ -1,264 +1,177 @@
 
 SERVICE_NAME = "InfectionMonkey"
-
-
-class CopiedFileDetails:
-    """Stores the details of a copied file"""
-
-    def __init__(self, remote_path: PurePath, destination_path: PurePath):
-        self.remote_path = remote_path
-        self.destination_path = destination_path
-
-
-class SMBExploitClient:
+LOGIN_TAGS = {
+    T1021_ATTACK_TECHNIQUE_TAG,  # Remote Services
+    T1110_ATTACK_TECHNIQUE_TAG,  # Brute Force
+    T1210_ATTACK_TECHNIQUE_TAG,  # Exploitation of Remote Services
+}
+SHARE_DISCOVERY_TAGS = {
+    T1135_ATTACK_TECHNIQUE_TAG,  # Network Share Discovery
+}
+COPY_FILE_TAGS = {
+    T1105_ATTACK_TECHNIQUE_TAG,  # Ingress Tool Transfer
+}
+EXECUTION_TAGS = {
+    T1021_ATTACK_TECHNIQUE_TAG,  # Remote Services
+    T1210_ATTACK_TECHNIQUE_TAG,  # Exploitation of Remote Services
+    T1569_ATTACK_TECHNIQUE_TAG,  # Execution: System Services
+}
+SMB_PORTS = [NetworkPort(139), NetworkPort(445)]
+
+
+class SMBExploitClient(IRemoteAccessClient):
     """Manages the SMB connection, Exploitation events"""
 
     def __init__(
-        self, agent_event_publisher: IAgentEventPublisher, host: TargetHost, options: SMBOptions
+        self,
+        host: TargetHost,
+        options: SMBOptions,
+        smb_client: SMBClient,
     ):
-        self._agent_event_publisher = agent_event_publisher
         self._host = host
         self._options = options
-        self._smb: Optional[SMBConnection] = None
-        self._copied_file_details: Optional[CopiedFileDetails] = None
-        self._authenticated_credentials: Optional[Credentials] = None
+        self._smb_client = smb_client
 
-    def authenticate(self, credentials: Credentials) -> bool:
-        """Returns True if authentication succeeded, False otherwise
-        Side effect: The SMB connection is established on success"""
+    def login(self, credentials: Credentials, tags: Set[str]):
+        tags.update(LOGIN_TAGS)
 
-        self._smb = create_smb_connection(self._host)
-        if not self._smb:
-            return False
-
-        if not smb_login(self._smb, credentials):
-            return False
-
-        self._smb.setTimeout(self._options.smb_connect_timeout)
-
-        if logout_guest(self._smb):
-            return False
+        try:
+            self._smb_client.connect_with_user(
+                self._host, credentials, timeout=self._options.smb_connect_timeout
+            )
+        except Exception as err:
+            error_message = f"Failed to authenticate over SMB with {credentials}: {err}"
+            raise RemoteAuthenticationError(error_message)
 
-        self._authenticated_credentials = credentials
-        return True
+    def _raise_if_not_authenticated(self, error_type: Type[Exception]):
+        if not self._smb_client.connected():
+            raise error_type(
+                "This operation cannot be performed until authentication is successful"
+            )
 
     def get_os(self) -> OperatingSystem:
         return OperatingSystem.WINDOWS
```

Comment:
Doesn't fulfill the interface as it doesn't query the remote system. Maybe we should reword the docstring in the interface